packages/snaps-controllers/src/snaps/location/npm.ts

```diff
@@ -0,0 +1,356 @@
+    const fetchFunction = opts.fetch ?? globalThis.fetch;
+    const requestedRange = opts.versionRange ?? DEFAULT_REQUESTED_SNAP_VERSION;
+
+    assertStruct(url.toString(), NpmSnapIdStruct, 'Invalid Snap Id: ');
+
+    let registry: string | URL;
+    if (
+      url.host === '' &&
+      url.port === '' &&
+      url.username === '' &&
+      url.password === ''
+    ) {
+      registry = new URL(DEFAULT_NPM_REGISTRY);
+    } else {
+      registry = 'https://';
+      if (url.username) {
+        registry += url.username;
+        if (url.password) {
+          registry += `:${url.password}`;
+        }
+        registry += '@';
+      }
+      registry += url.host;
+      registry = new URL(registry);
+      assert(
+        allowCustomRegistries,
+        new TypeError(
+          `Custom NPM registries are disabled, tried to use "${registry.toString()}".`,
+        ),
+      );
+    }
+
+    assert(
+      registry.pathname === '/' &&
+        registry.search === '' &&
+        registry.hash === '',
+    );
+
+    assert(
+      url.pathname !== '' && url.pathname !== '/',
+      new TypeError('The package name in NPM location is empty.'),
+    );
+    let packageName = url.pathname;
+    if (packageName.startsWith('/')) {
+      packageName = packageName.slice(1);
+    }
+
+    this.meta = {
+      requestedRange,
+      registry,
+      packageName,
+      fetch: fetchFunction,
+    };
+  }
+
+  async manifest(): Promise<VirtualFile<SnapManifest>> {
+    if (this.validatedManifest) {
+      return this.validatedManifest.clone();
+    }
+
+    const vfile = await this.fetch('./snap.manifest.json');
+    const result = JSON.parse(vfile.toString());
+    assertIsSnapManifest(result);
+    vfile.result = result;
+    this.validatedManifest = vfile as VirtualFile<SnapManifest>;
+
+    return this.manifest();
+  }
+
+  async fetch(path: string): Promise<VirtualFile> {
+    const relativePath = ensureRelative(path);
+    if (!this.files) {
+      await this.#lazyInit();
+      assert(this.files !== undefined);
+    }
+    const vfile = this.files.get(relativePath);
+    assert(
+      vfile !== undefined,
+      new TypeError(`File "${path}" not found in package.`),
+    );
+    return vfile.clone();
+  }
+
+  get packageName(): string {
+    return this.meta.packageName;
+  }
+
+  get version(): string {
+    assert(
+      this.meta.version !== undefined,
+      'Tried to access version without first fetching NPM package.',
+    );
+    return this.meta.version;
+  }
+
+  get registry(): URL {
+    return this.meta.registry;
+  }
+
+  get versionRange(): SemVerRange {
+    return this.meta.requestedRange;
+  }
+
+  async #lazyInit() {
+    assert(this.files === undefined);
+    const [tarballResponse, actualVersion] = await fetchNpmTarball(
+      this.meta.packageName,
+      this.meta.requestedRange,
+      this.meta.registry,
+      this.meta.fetch,
+    );
+    this.meta.version = actualVersion;
+
+    let canonicalBase = 'npm://';
```

Comment:
Should this be `npm:` similarly to how local canonical paths start with `local:`?

Reply:
I wanted npm (and hopefully all snap locations in the future) paths to follow SIP-8 and URI design.

`URI = scheme ":" ["//" authority] path ["?" query] ["#" fragment]`

* `authority := the registry`
* `path := package name`

So it'll be `npm://registry.npmjs.org/my-package` in full canonical form. `npm:my-package` in shorthand canonical form.


We have a default registry, but that not always was a case - we used mirrors previously. That's I want to always use the full form internally.